src/algebra/order/monoid_lemmas_gt_zero.lean

```diff
@@ -0,0 +1,131 @@
+Authors: Damiano Testa
+-/
+import algebra.covariant_and_contravariant
+
+/-!
+# Multiplication by ·positive· elements is monotonic
+
+Let `α` be a type with `<` and `0`.  We introduce the type `gt_zero α = {x : α // 0 < x}` of
+positive elements of `α`.  We also introduce a local notation for the subtype `gt_zero α`:
+
+*  the notation `α>0` to stands for `gt_zero α`.
+
+If the type `α` also has a multiplication, then we also define the multiplications on the left and
+on the right of an element of `α>0` and an element of `α`:
+
+*  `sx : α>0 → α → α` is defined as `sx a b = a * b`, with `a` coerced to `α` by virtue of being in
+  a subtype of `α`;
+*  `dx : α>0 → α → α` is defined as `dx a b = b * a`, with `a` coerced to `α` by virtue of being in
+  a subtype of `α`.
+
+We combine this with (`contravariant_`) `covariant_class`es to assume that multiplication by
+positive elements is (strictly) monotone on a `mul_zero_class`, `monoid_with_zero`,...
+More specifically, we use extensively the following typeclasses:
+
+* monotone left
+* * `covariant_class α>0 α sx (≤)`, expressing that multiplication by positive elements on the left
+    is monotone;
+* * `covariant_class α>0 α sx (<)`, expressing that multiplication by positive elements on the left
+    is strictly monotone;
+* monotone right
+* * `covariant_class α>0 α dx (≤)`, expressing that multiplication by positive elements on the right
+    is monotone;
+* * `covariant_class α>0 α dx (<)`, expressing that multiplication by positive elements on the right
+    is strictly monotone.
+* reverse monotone left
+* * `contravariant_class α>0 α sx (≤)`, expressing that multiplication by positive elements on the
+    left is reverse monotone;
+* * `contravariant_class α>0 α sx (<)`, expressing that multiplication by positive elements on the
+    left is strictly reverse monotone;
+* reverse reverse monotone right
+* * `contravariant_class α>0 α dx (≤)`, expressing that multiplication by positive elements on the
+    right is reverse monotone;
+* * `contravariant_class α>0 α dx (<)`, expressing that multiplication by positive elements on the
+    right is strictly reverse monotone.
+
+##  Formalization comments
+
+We use `α>0 = {x : α // 0 < x}` with a strict inequality since in most cases what happens with `0`
+is clear.  This creates a few bumps in the first couple of proofs, where we have to split cases on
+whether an element is `0` or not, but goes smoothly after that.  A further advantage is that we
+only introduce notation for the positive elements and we do not need also the non-negative ones.
+-/
+
+/- I am changing the file `algebra/order/monoid_lemmas` incrementally, with the idea of
+reproducing almost all of the proofs in `algebra/order/ring` with weaker assumptions. -/
+
+universe u
+variable {α : Type u}
+
+/--  The type of positive elements in a type that has a `0` and a `<`. -/
+@[reducible]
+def gt_zero (α : Type u) [has_zero α] [has_lt α] := {x : α // 0 < x}
```

Comment:
I think Floris did the `nonneg` analogue without introducing a new definition. Would that also work here?

Reply:
Nitpick in the case a new definition is needed: I'd rather see `zero_lt` than `gt_zero`, to reflect the actual contents of the definition.

Reply:
If the notation works without introducing a new definition, I would also be in favor of not adding this definition.

Reply:
I'm still happy if the declarations below are in some namespace

Reply:
Removing the definition causes no problem to the local notation: I removed the definition.

I left the namespace and have changed its name to `zero_lt`.